regression-test/src/androidTest/java/okhttp/regression/keys/ClientAuthAndroidTest.java

```diff
@@ -0,0 +1,175 @@
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public String chooseServerAlias(String keyType, Principal[] issuers, Socket socket) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public X509Certificate[] getCertificateChain(String alias) {
+            return chain;
+        }
+
+        @Override
+        public PrivateKey getPrivateKey(String alias) {
+            return pk;
+        }
+    }
+
+    @Test
+    public void generateCertificate() throws IOException, NoSuchAlgorithmException, KeyManagementException {
+        HandshakeCertificates m = new HandshakeCertificates.Builder()
+                .addPlatformTrustedCertificates()
+                .build();
+
+//        java.lang.SecurityException: exception: java.security.NoSuchAlgorithmException: The BC provider no longer
+//        provides an implementation for Signature.SHA256withECDSA.  Please see
+//        https://android-developers.googleblog.com/2018/03/cryptography-changes-in-android-p.html for more details.
+//        at org.bouncycastle.x509.X509V3CertificateGenerator.generateX509Certificate(Unknown Source:25)
+//        at org.bouncycastle.x509.X509V3CertificateGenerator.generateX509Certificate(Unknown Source:3)
+        HeldCertificate localhost = new HeldCertificate.Builder()
+                .commonName("localhost")
+                .addSubjectAlternativeName(InetAddress.getByName("localhost").getCanonicalHostName())
+                .build();
+
+        FixedKeyManager keyManager = new FixedKeyManager(localhost.keyPair().getPrivate(), localhost.certificate());
+
+        SSLContext sslContext = SSLContext.getInstance("TLS");
+        sslContext.init(new KeyManager[]{keyManager}, new TrustManager[]{m.trustManager()}, null);
+
+        OkHttpClient client = new OkHttpClient.Builder()
+                .sslSocketFactory(sslContext.getSocketFactory(), m.trustManager())
+                .build();
+
+        Request request = new Request.Builder()
+                .url("https://server.cryptomix.com/secure/")
+                .build();
+        try (Response response = client.newCall(request).execute()) {
+            assertEquals(200, response.code());
+
+            System.out.println(response.body().string());
+        }
+    }
+
+    @Test
+    public void keyStoreCertificate() throws IOException, NoSuchAlgorithmException, KeyManagementException, KeyStoreException, KeyChainException, InterruptedException {
+        TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+        trustManagerFactory.init((KeyStore) null);
+        X509TrustManager trustManager = (X509TrustManager) trustManagerFactory.getTrustManagers()[0];
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        String[] selectedAlias = new String[1];
+        KeyChain.choosePrivateKeyAlias(null, // TODO launch with an activity
```

Comment:
Fails here as I'll need to launch an activity to test at least first time.  Maybe some permission is available to allow managing keys and certs.